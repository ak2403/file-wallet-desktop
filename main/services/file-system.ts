import fs from 'fs';
import mime from 'mime';

import { FileInformation } from '../../packages/types/src/services';

export class FileSystemService {
  static getFileInfo(filePath: string): FileInformation {
    try {
      const isFileExists = fs.existsSync(filePath);

      if (!isFileExists) {
        return {
          exists: false,
        };
      }

      const fileStats = fs.statSync(filePath);

      return {
        exists: true,
        stats: {
          filePath,
          fileType: mime.getType(filePath),
          size: fileStats.size,
        },
      };
    } catch (error) {
      throw new Error(error);
    }
  }
}
