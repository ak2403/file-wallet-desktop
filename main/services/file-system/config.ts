export const RootUserName = `/Users/${process.env['LOGNAME']}`;
