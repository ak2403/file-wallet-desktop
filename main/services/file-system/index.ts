export * from './read-file';
export * from './read-folder';
export * from './write-file';
