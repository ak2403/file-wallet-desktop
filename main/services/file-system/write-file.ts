import fs from 'fs';

export const writeFile = async (path: string, data: Buffer): Promise<boolean> => {
  try {
    await fs.writeFileSync(path, data, {
      encoding: 'base64',
    });

    return true;
  } catch (err) {
    console.error(err);

    return false;
  }
};
