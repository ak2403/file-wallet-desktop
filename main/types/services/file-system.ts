export type ReadFile = {
  type: string;
  buffer: Buffer;
};

export enum StructureType {
  Folder = 'folder',
  File = 'file',
}

export type FolderStructure = {
  id: string;
  name: string;
  ext: string;
  type: StructureType;
};
