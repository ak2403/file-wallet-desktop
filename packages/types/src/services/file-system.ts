export type FileInformation = {
  exists: boolean;
  stats?: {
    fileType: string | null;
    size: number;
  };
};

export type DialogFileInformation = FileInformation & {
  canceled: boolean;
  errored?: boolean;
};
