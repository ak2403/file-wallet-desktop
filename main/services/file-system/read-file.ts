import fs from 'fs';
import { extname } from 'path';

import { RootUserName } from './config';
import { ReadFile } from '../../types/services';

export const readFile = async (filePath: string = ''): Promise<ReadFile | null> => {
  try {
    const fileType = extname(`${RootUserName}/${filePath}`);

    const fileData = await fs.readFileSync(`${RootUserName}/${filePath}`);

    const fileBuffer = Buffer.from(fileData);

    return {
      type: fileType,
      buffer: fileBuffer,
    };
  } catch (err) {
    console.log(err);
    return null;
  }
};
