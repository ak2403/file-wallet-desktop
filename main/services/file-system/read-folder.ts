import fs from 'fs';
import { extname } from 'path';
import { v4 as uuid } from 'uuid';

import { RootUserName } from './config';
import { FolderStructure, StructureType } from '../../types/services';

export const readFolder = async (path: string = ''): Promise<FolderStructure[]> => {
  try {
    const folderInformation = await fs.readdirSync(`${RootUserName}/${path}`);
    const folderDirectories: FolderStructure[] = [];

    for (const file of folderInformation) {
      const fileStat = await fs.statSync(`${RootUserName}/${path}/${file}`, {
        bigint: true,
      });

      if (file && file[0] !== '.') {
        folderDirectories.push({
          id: uuid(),
          name: file,
          ext: fileStat.isDirectory() ? 'folder' : extname(file),
          type: fileStat.isDirectory() ? StructureType.Folder : StructureType.File,
        });
      }
    }

    return folderDirectories.sort((a, b) => b.type.localeCompare(a.type) || a.name.localeCompare(b.name));
  } catch (err) {
    console.log(err);

    return [];
  }
};
