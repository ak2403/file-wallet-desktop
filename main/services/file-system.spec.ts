import path from 'path';
import { FileSystemService } from './file-system';

describe('FileSystemService', () => {
  describe('getFileInfo()', () => {
    it('returns file information for valid file path', () => {
      const filePath = path.join(__dirname, './file-system.ts');

      const response = FileSystemService.getFileInfo(filePath);

      expect(response).toMatchSnapshot();
    });

    it('returns with exists false when path invalid', () => {
      const filePath = path.join(__dirname, './dummy_path.txt');

      const response = FileSystemService.getFileInfo(filePath);

      expect(response.exists).toBeFalsy();
    });
  });
});
