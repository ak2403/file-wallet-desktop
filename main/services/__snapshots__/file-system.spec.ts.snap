// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSystemService getFileInfo() returns file information for valid file path 1`] = `
{
  "exists": true,
  "stats": {
    "fileType": "video/mp2t",
    "size": 622,
  },
}
`;
